### PR TITLE
CMD: Detect and add game(s) with commandline, Feature request #7885

### PR DESCRIPTION
**Pull request by yinsimei**
GCoS 2017. Move the cmd commit to a branch to work in master without affecting the PR.

### CODE REVIEW

File: base/commandLine.cpp [hunk 1]
```diff
@@ -68,6 +69,13 @@ static const char HELP_STRING[] =
 	"  -z, --list-games         Display list of supported games and exit\n"
 	"  -t, --list-targets       Display list of configured targets and exit\n"
 	"  --list-saves=TARGET      Display a list of saved games for the game (TARGET) specified\n"
+	"  -a, --add                Add a game from current or specified directory\n"
+	"                           Use --path=PATH before -a, --add to specify a directory.\n"
+	"  --massadd                Add all games from current or specified directory and all sub directories\n"
+	"                           Use --path=PATH before --massadd to specify a directory.\n"
+	"  -d, --detect             Display a list of games from current or specified directory\n"
```
Comment (criezy) [inline base/commandLine.cpp]: -d is already used for the debug level

File: base/commandLine.cpp [hunk 2]
```diff
@@ -798,20 +815,139 @@ static bool autoDetect(Common::String path) {
...
+		return false;
+
+	int idx = 0;
+	// Pick one if there are several games
+	if (candidates.size() > 1) {
+		// Print game list
+		printf("Several games are detected. Please pick one game to add: \n");
+		int i = 1;
+		for (GameList::iterator v = candidates.begin(); v != candidates.end(); ++i, ++v) {
+			printf("%2i. %s\n", i, v->description().c_str());
```
Comment (criezy) [inline base/commandLine.cpp]: Maybe print the gameid as well, for consistency with detectGames(), and because some users are used to those.

File: base/commandLine.cpp [hunk 3]
```diff
@@ -798,20 +815,139 @@ static bool autoDetect(Common::String path) {
...
+			for (GameList::iterator v = candidates.begin(); v != candidates.end(); ++v) {
+
+				++ndetect;
+				Common::String domain = v->preferredtarget();
+				// If game was already added, don't add it
+				if (ConfMan.hasGameDomain(domain))
+					continue;
+
+				// Add game
+				ConfMan.addGameDomain(domain);
```
Comment (criezy) [inline base/commandLine.cpp]: This is not sufficient, is it? Have you tested that you can actually start the game from the launcher after adding it this way. I would expect that at least the path would need to be added to ConfMan as well for the new domain.

File: base/commandLine.cpp [hunk 4]
```diff
@@ -798,20 +815,139 @@ static bool autoDetect(Common::String path) {
...
+			candidates[idx].gameid().c_str(),
+			candidates[idx].description().c_str(),
+			Common::getLanguageDescription(candidates[idx].language()),
+			Common::getPlatformDescription(candidates[idx].platform()));
+	return true;
+}
+
+static bool massAddGame(Common::String path) {
+	if (path.empty())
+			path = ".";
```
Comment (criezy) [inline base/commandLine.cpp]: Indentation is wrong for this line.

File: base/commandLine.cpp [hunk 5]
```diff
@@ -1039,8 +1175,15 @@ bool processSettings(Common::String &command, Common::StringMap &settings, Commo
...
 	} else if (command == "auto-detect") {
 		// If auto-detects succeed, we want to return false so that the game is started
-		return !autoDetect(settings["path"]);
+		return !detectGames(settings["path"], true);
 		//return true;
+	} else if (command == "detect") {
+		detectGames(settings["path"], false);
+		return true;
+	} else if (command == "add") {
+		return !addGame(settings["path"]);
```
Comment (criezy) [inline base/commandLine.cpp]: Do we actually want to start ScummVM after adding a game with the command line?
The same question also applies to mass add below.

I don't have a definite answer to this question. But I think I would not expect ScummVM to start after adding the game(s).

**Comment by criezy** [comment]
Overall this looks quite good, although there are some changes that need to be made.

I am also wondering if we could avoid some code duplication between the GUI and the command line for adding games to the config file and for the mass add detection code.

File: base/commandLine.cpp (hunk 1 above)
Comment (yinsimei) [inline base/commandLine.cpp]: Ok, I will remove it

File: base/commandLine.cpp (hunk 2 above)
Comment (yinsimei) [inline base/commandLine.cpp]: Ok, I will add it.

File: base/commandLine.cpp (hunk 3 above)
Comment (yinsimei) [inline base/commandLine.cpp]: Yes, I've tested it. The game can be started from launcher. I think the path has already been taken care of in the function detectGames of metaengine or somewhere else. I will take a deeper look into how it works exactly.

File: base/commandLine.cpp (hunk 5 above)
Comment (yinsimei) [inline base/commandLine.cpp]: Yes, it's not specified in the feature request. I will change it.

Comment (wjp) [inline base/commandLine.cpp]: For context, feature requests are just that: requests. They haven't necessarily been discussed or fully worked out.

**Comment by wjp** [comment]
"Feature request #703" is old SF/Allura numbering. You should mention 7885 instead ( https://bugs.scummvm.org/ticket/7885 ).

**Comment by yinsimei** [comment]
> I am also wondering if we could avoid some code duplication between the GUI and the command line for adding games to the config file and for the mass add detection code.

Yes, I've noticed the duplication. Do think it will be a good idea to have a massDetectGames in MetaEngine?

File: base/commandLine.cpp (hunk 3 above)
Comment (yinsimei) [inline base/commandLine.cpp]: No, you are right, it's not sufficient. The game can only be started at the first time and then there is a problem after re-opening the launcher. I will look into it

**Comment by criezy** [comment]
Thank you.
I have now merged this change after squashing the two commits.

File: base/commandLine.cpp [hunk 6]
```diff
@@ -68,6 +69,13 @@ static const char HELP_STRING[] =
 	"  -z, --list-games         Display list of supported games and exit\n"
 	"  -t, --list-targets       Display list of configured targets and exit\n"
 	"  --list-saves=TARGET      Display a list of saved games for the game (TARGET) specified\n"
+	"  -a, --add                Add a game from current or specified directory\n"
+	"                           Use --path=PATH before -a, --add to specify a directory.\n"
+	"  --massadd                Add all games from current or specified directory and all sub directories\n"
```
Comment (tobiatesan) [inline base/commandLine.cpp]: This is very useful, but I find --massadd is super unusual as a command line switch. Could we change this to `--all` (perhaps with `-A`) or something before 1.10?

@yinsimei @criezy @sev-

Comment (tobiatesan) [inline base/commandLine.cpp]: We've discussed the issue a bit with @criezy, see:

TL;DR, I don't find this very intuitive and I don't especially like the idea of interaction (if you want interaction, you'd use a GUI after all...)

The semantics I'd like to propose is:

```
-p <dir> --add                adds all games
-p <dir> --detect             enumerates dectected games in <dir>
-p <dir> --game <id> --add     adds just game <id>
```

How do you like it?

Comment (tobiatesan) [inline base/commandLine.cpp]: (If you like the current semantics, I think the help should at least mention that `-a` is interactive -- especially since it's interactive only some of the time; that might give an headache to some script writer and/or cabinet modder.)

Comment (bgK) [inline base/commandLine.cpp]: That sounds good and actually usable by third party launchers.

Maybe the `addGameToConf` function could additionally display the target of added games so that it can be used for subsequent calls to scummvm.

Comment (yinsimei) [inline base/commandLine.cpp]: So, if I understand well, we'll use --add in the place of --massadd and add a --game option for --add ?

Comment (tobiatesan) [inline base/commandLine.cpp]: @yinsimei yes, that's essentially the idea. And `--add` would error out if there are multiple games with no `--game` option specified (possibly enumerating them or just advising to use `--detect`)
@bgK good idea, I like it.

Comment (criezy) [inline base/commandLine.cpp]: > And --add would error out if there are multiple games and no --game option is passed

@tobiatesan I though we agreed during the discussion on IRC today that --add without --game would add all the games found (like the current --massadd)? And that is also what one of your earlier comment above says.

Comment (tobiatesan) [inline base/commandLine.cpp]: @criezy yes, sorry, that's correct. I honestly have no idea why I wrote the bit you quoted.
